sentry/src/middleware/auth.rs

```diff
@@ -0,0 +1,95 @@
+use hyper::header::AUTHORIZATION;
+use hyper::{Body, Request};
+use redis::aio::SharedConnection;
+
+use primitives::adapter::{Adapter, Session as AdapterSession};
+
+use crate::{ResponseError, Session};
+
+/// Check `Authorization` header for `Bearer` scheme with `Adapter::session_from_token`.
+/// If the `Adapter` fails to create an `AdapterSession`, `ResponseError::BadRequest` will be returned.
+pub(crate) async fn for_request(
+    mut req: Request<Body>,
+    adapter: impl Adapter,
+    redis: SharedConnection,
+) -> Result<Request<Body>, ResponseError> {
+    let authorization = req.headers().get(AUTHORIZATION);
+
+    let prefix = "Bearer ";
+
+    let token = authorization
+        .and_then(|hv| {
+            hv.to_str()
+                .map(|token_str| {
+                    if token_str.starts_with(prefix) {
+                        None
+                    } else {
+                        Some(token_str.replacen(prefix, "", 1))
+                    }
+                })
+                .transpose()
+        })
+        .transpose()?;
+
+    if let Some(ref token) = token {
+        let adapter_session = match redis::cmd("GET")
+            .arg(token)
+            .query_async::<_, Option<String>>(&mut redis.clone())
+            .await?
+        {
+            Some(session) => {
+                match serde_json::from_str::<AdapterSession>(&session) {
+                    Ok(session) => session,
+                    Err(serde_error) => {
+                        // On a Deserialization error remove the key from Redis as a precaution
+                        // DEL will return the number of deleted keys, but we don't need them
+                        let _ = redis::cmd("DEL")
```

Comment:
just in case we have a "wrong" value in `Redis`, we are deleting it as a precaution.
Do you think this is necessary @Ivshti @samparsky ?

Reply:
I don't think this is necessary Redis would overwrite values with the same key with the newest value

Reply:
It can be, because if it errors after the deserialization you won't get to the `SET`...

Reply:
This is resolved. If there is a wrong value in Redis, we will try to fetch the token again with the adapter and override the Redis value with it.